subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/MapTypeHandlerFactory.java

```diff
@@ -50,6 +48,19 @@
                 context
         );
 
-        return Optional.of((TypeHandler<T>) new StringMapTypeHandler<>(valueTypeHandler));
+        if (String.class.equals(keyType)) {
+            return Optional.of((TypeHandler<T>) new StringMapTypeHandler<>(valueTypeHandler));
+        } else {
+            Optional<TypeHandler<?>> declaredKeyTypeHandler =
+                    context.getTypeHandlerLibrary().getTypeHandler(keyType);
+            TypeInfo<?> keyTypeInfo = TypeInfo.of(keyType);
+            @SuppressWarnings({"unchecked"})
+            TypeHandler<?> keyTypeHandler = new RuntimeDelegatingTypeHandler(
+                    declaredKeyTypeHandler.orElse(null),
+                    keyTypeInfo,
```

Comment:
I'm wondering about the SuppressWarnings here. I guess it was probably copied from valueTypeHandler above.

Reply:
I guess we cannot get type safety for the generics in here 🤷‍♂️